File: src/main/java/br/com/dusty/dkits/listener/mechanics/AsyncPlayerChatListener.java
package br.com.dusty.dkits.listener.mechanics;

import br.com.dusty.dkits.gamer.EnumChat;
import br.com.dusty.dkits.gamer.EnumRank;
import br.com.dusty.dkits.gamer.Gamer;
import br.com.dusty.dkits.gamer.GamerRegistry;
import br.com.dusty.dkits.util.text.Text;
import br.com.dusty.dkits.util.text.TextStyle;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

public class AsyncPlayerChatListener implements Listener {
	
	private static final Text STAFF_CHAT_PREFIX_NEUTRAL = Text.basicOf("[").neutral("Staff Chat").basic("] ");
	private static final Text STAFF_CHAT_PREFIX_NEGATIVE = Text.basicOf("[").neutral("Staff Chat").basic("] ");
	
	@EventHandler
	public void onAsyncPlayerChat(AsyncPlayerChatEvent event) {
		Player player = event.getPlayer();
		
		Gamer gamer = Gamer.of(player);
		
		switch(gamer.getChat()){
			case NORMAL:
				event.setFormat("<%s" + TextStyle.RESET + "> %s");
				
				break;
			case STAFF:
				event.getRecipients().clear();
				
				String message_neutral = STAFF_CHAT_PREFIX_NEUTRAL.append("<")
				                                                  .append(player.getDisplayName())
				                                                  .append("> ")
				                                                  .neutral(event.getMessage())
				                                                  .toString();
				
				String message_negative = STAFF_CHAT_PREFIX_NEGATIVE.append("<")
				                                                    .append(player.getDisplayName())
				                                                    .append("> ")
				                                                    .negative(event.getMessage())
				                                                    .toString();
				
				for(Gamer otherGamer : GamerRegistry.getOnlineGamers())
					if(otherGamer.getRank().isHigherThanOrEquals(EnumRank.MOD))
						otherGamer.getPlayer()
						          .sendMessage(otherGamer.getChat() == EnumChat.STAFF ? message_neutral : message_negative);
				
				break;
		}
	}
}

File: src/main/kotlin/br/com/dusty/dkits/listener/mechanics/AsyncPlayerChatListener.kt
package br.com.dusty.dkits.listener.mechanics

import br.com.dusty.dkits.gamer.EnumChat.*
import br.com.dusty.dkits.gamer.EnumRank
import br.com.dusty.dkits.gamer.GamerRegistry
import br.com.dusty.dkits.util.gamer.gamer
import br.com.dusty.dkits.util.text.Text
import br.com.dusty.dkits.util.text.TextColor
import br.com.dusty.dkits.util.text.TextStyle
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent

object AsyncPlayerChatListener: Listener {

	val STAFF_CHAT_PREFIX_POSITIVE = Text.basicOf("[").positive("Staff Chat").basic("] ")
	val STAFF_CHAT_PREFIX_NEGATIVE = Text.basicOf("[").negative("Staff Chat").basic("] ")

	val CLAN_CHAT_PREFIX_POSITIVE = Text.basicOf("[").positive("Clan Chat").basic("] ")
	val CLAN_CHAT_PREFIX_NEGATIVE = Text.basicOf("[").negative("Clan Chat").basic("] ")

	var rank = EnumRank.DEFAULT

	@EventHandler
	fun onAsyncPlayerChat(event: AsyncPlayerChatEvent) {
		val player = event.player
		val gamer = player.gamer()

		if (gamer.warp.overrides(event)) return

		when (gamer.chat) {
			NORMAL -> {
				if (gamer.rank.isLowerThan(rank)) {
					event.isCancelled = true

					player.sendMessage(Text.negativePrefix().basic("O chat está ").negative("restrito").basic(" apenas a ").negative(rank.string).basic(" e acima!").toString())
				}

				event.format = "" + Text.basicOf(if (gamer.clan != null) gamer.clan!!.tag + " " else "").toString() + "%s" + TextColor.GRAY + ":" + TextStyle.RESET + " %s"
			}
			CLAN   -> {
				event.recipients.clear()

				val messagePositive = CLAN_CHAT_PREFIX_POSITIVE.basic(if (gamer.clan != null) gamer.clan!!.tag + " " else "").append(gamer.rank.format(player.name) + TextStyle.RESET).basic(": ").neutral(
						event.message).toString()
				val messageNegative = CLAN_CHAT_PREFIX_NEGATIVE.basic(if (gamer.clan != null) gamer.clan!!.tag + " " else "").append(gamer.rank.format(player.name) + TextStyle.RESET).basic(": ").negative(
						event.message).toString()

				GamerRegistry.onlineGamers().filter { it.clan == gamer.clan }.forEach {
					it.player.sendMessage(if (it.chat == CLAN) messagePositive else messageNegative)
				}
			}
			STAFF  -> {
				event.recipients.clear()

				val messagePositive = STAFF_CHAT_PREFIX_POSITIVE.basic(if (gamer.clan != null) gamer.clan!!.tag + " " else "").append(gamer.rank.format(player.name) + TextStyle.RESET).basic(": ").neutral(
						event.message).toString()
				val messageNegative = STAFF_CHAT_PREFIX_NEGATIVE.basic(if (gamer.clan != null) gamer.clan!!.tag + " " else "").append(gamer.rank.format(player.name) + TextStyle.RESET).basic(": ").negative(
						event.message).toString()

				GamerRegistry.onlineGamers().filter { it.rank.isHigherThanOrEquals(EnumRank.MOD) }.forEach {
					it.player.sendMessage(if (it.chat == STAFF) messagePositive else messageNegative)
				}
			}
		}
	}
}
